src/main/kotlin/uk/gov/justice/digital/hmpps/manageoffencesapi/service/OffenceService.kt

```diff
@@ -23,6 +30,125 @@ class OffenceService(
     return sdrsLoadResultRepository.findAllByOrderByAlphaCharAsc().map { transform(it) }
   }
 
+  @Transactional
+  fun fullSyncWithNomis() {
+    ('A'..'Z').forEach { alphaChar ->
+      SDRSService.log.info("Starting full sync with NOMIS for alphaChar {} ", alphaChar)
+      fullySyncOffenceGroupWithNomis(alphaChar.toString())
+    }
+  }
+
+  private fun fullySyncOffenceGroupWithNomis(alphaChar: String) {
+    val offencesByCode = offenceRepository.findByCodeStartsWithIgnoreCase(alphaChar).associateBy { it.code }
+    val (nomisOffencesById, nomisOffences) = getAllNomisOffencecsForAlphaChar(alphaChar)
+
+    // the keys here represent the NOMIS keys, each key is a pair of the offenceCode and the statuteCode
+    val (existingOffenceKeys, newOffenceKeys) = offencesByCode.keys.map { Pair(it, it.substring(0, 4)) }
```

Comment:
There's quite a bit of `substring(0, 4)` going on - might be more readable if you create an extension function, e.g. `private fun String.statuteCode() = this.substring(0, 4)`

Reply:
Good spot, the first four characters of the offence code is the statute code. We are planning to store that in a separate colum (therefore extract it when we initially create an offence). So all this will go away when we split out the statute into its own separate field